### PR TITLE
Protocol v11

### PR DESCRIPTION
**Pull request by WhoSoup**
This is a draft for the potential V11 protocol. It includes significant changes to controller_connections and the handshake. 

The handshake procedure is reworked from being asymmetrical to being symmetrical. The dialing node will send their handshake first, then listen to a reply. The listening node will listen for a handshake first, then send a reply. This is the only way for the listening node to be able to upgrade the handshake to protocol V11 for third-party apps that only understand protobuf.

See the README for a detailed description of the handshake process.

The detection between Gob and Protobuf is accomplished by sending a specific 4-byte-sequence (0xfafafafa) first using a buffered Peek(). If the sequence is present, the buffer is dropped and it continues with the (size, message) Protobuf tuples. Otherwise, it initiates a gob-decoder on the buffer.

There is some complication with V9 and this implementation. For the outgoing handshake, it creates two gob encoder instances, once for the desired protocol ("ours"), once for the detected protocol ("theirs). On the incoming handshake side, it only creates one gob decoder instance. Gob is stateful and will break if the same type is registered twice in a row. There are two solutions:
1. Create a new gob decoder on the incoming side after the handshake is read
2. Re-use the first gob encoder on the outgoing side

I went with option 2. 

This problem is not present in V10 since the gob state switches from the V9 Handshake to the V10 message. 

TODO:
* Clean up leftover debug code
* Reorganize the handshakes to de-duplicate code and make it more readable
* Test the new code on a local net and the factom testnet for compatibility
* (optional) Verify a third party app that only understands protobuf can connect to a node

### CODE REVIEW

**Comment by WhoSoup** [comment]
Managed to track down an annoying bug and the V11 code is now capable of connecting to the test network and sync without errors.

File: endpoint.go [hunk 1]
```diff
@@ -2,16 +2,27 @@ package p2p
...
-//                     (      optional labels        )    (    required label            )
+func testRandomEndpoint() Endpoint {
+	ip := fmt.Sprintf("%d.%d.%d.%d", 1+rand.Intn(255), rand.Intn(256), rand.Intn(256), 1+rand.Intn(255))
+	return Endpoint{IP: ip, Port: testRandomPort()}
+}
+
+func testRandomPort() string {
+	return fmt.Sprintf("%d", (1+rand.Uint32())%math.MaxUint16)
+}
+
```
Comment (PaulBernier) [inline endpoint.go]: shouldn't those rather be in `endpoint_test.go` file?

File: prometheus.go [hunk 2]
```diff
@@ -43,18 +52,32 @@ func (p *Prometheus) Setup() {
...
 		p.AppReceived = ng("factomd_p2p_messages_received", "Total number of application messages received")
-		p.AppDuplicate = ng("factomd_p2p_messages_duplicate", "Total number of duplicate messages filtered out")
+		p.ByteRateDown = ng("factomd_p2p_byte_rate_down", "Current rate of incoming traffic (in bytes/sec)")
+		p.ByteRateUp = ng("factomd_p2p_byte_rate_up", "Current rate of outgoing traffic (in bytes/sec)")
+		p.MessageRateDown = ng("factomd_p2p_msg_rate_down", "Current rate of incoming traffic (in messages/sec)")
+		p.MessageRateUp = ng("factomd_p2p_msg_rate_up", "Current rate of outgoing traffic (in messages/sec)")
+
+		p.DroppedPeerSend = ng("factomd_p2p_dropped_peer_send", "Number of parcels that were unable to be sent because of peer's lack of bandwidth")
+		p.DroppedFromNetwork = ng("factomd_p2p_dropped_from_network", "Number of parcels that were dropped because application is sending parcels too fast")
+		p.DroppedToNetwork = ng("factomd_p2p_dropped_to_network", "Number of parcels that were dropped because application is not reading parcels fast enough")
```
Comment (PaulBernier) [inline prometheus.go]: Are the description of those two gauges swapped?
FromNetwork is when we receiving a message *from* the network and we forward to the consuming application for reading.

File: protocolV10.go [hunk 3]
```diff
@@ -19,26 +18,42 @@ type ProtocolV10 struct {
 // V10Msg is the barebone message
 type V10Msg struct {
 	Type    ParcelType
-	Crc32   uint32
 	Payload []byte
 }
 
-func (v10 *ProtocolV10) init(decoder *gob.Decoder, encoder *gob.Encoder) {
+func newProtocolV10(decoder *gob.Decoder, encoder *gob.Encoder) *ProtocolV10 {
```
Comment (PaulBernier) [inline protocolV10.go:24]: Given than v10 was never deployed live, could we just remove all the code assiociated all together?

File: README.md [hunk 4]
```diff
@@ -105,29 +105,57 @@ Peers that are rejected are given a list of 3 (conf: `PeerShareAmount`) random p
...
 Data is serialized via Golang's gob.
 
 ### 10
 
-Protocol 10 is the slimmed down version of V9, containing only the Type, CRC32 of the payload, and the payload itself. Data is also serialized via Golang's gob.
+Protocol 10 is the slimmed down version of V9, containing only the Type, CRC32 of the payload, and the payload itself. Data is also serialized via Golang's gob. The handshake is encoded using V9's format.
+
+### 11
+
+Protocol 11 uses Protobuf ([protocolV11.proto](protocolV11.proto)) to define the Handshake and message. To signal that the connection is used for V11, the 4-byte sequence `0xfafafafa` is transmitted first. Protobufs are transmitted by sending the size of the marshalled protobuf first, encoded as uint32 in Big Endian format, followed by the protobuf byte sequence itself.
```
Comment (PaulBernier) [inline README.md]: 4 byte sequence doesn't match actual implementation

File: controller_cat.go [hunk 5]
```diff
@@ -14,8 +14,13 @@ func (c *controller) runCatRound() {
 	c.lastRound = time.Now()
 	c.logger.Debug("Cat Round")
 	c.rounds++
+	if c.net.prom != nil {
+		c.net.prom.CatRounds.Inc()
+	}
 
-	c.persistPeerFile()
+	if err := c.writePeerCache(); err != nil {
+		c.logger.Errorf("unable to write peer cache to disk: %v", err)
```
Comment (PaulBernier) [inline controller_cat.go]: this is loggrus isn't? So should probably be logged with `WithError()`

File: endpoint.go (hunk 1 above)
Comment (WhoSoup) [reply endpoint.go]: yes!

File: prometheus.go (hunk 2 above)
Comment (WhoSoup) [reply prometheus.go]: It would appear so

File: protocolV10.go (hunk 3 above)
Comment (WhoSoup) [reply protocolV10.go:24]: No, V10 is still a useful middlestep. Nodes that are configured to run V11 as preferred protocol are not backward compatible with V9, so without V10 we would still have to use V9 for all nodes.

V10 offers a significant reduction in bandwidth and it also lets us easily tell what % of nodes have upgraded.

**Comment by WhoSoup** [comment]
Merged in #42 to bring unit test coverage up to 59%

**Comment by WhoSoup** [comment]
Test coverage now up to 71.5%. Planning on merging this on this Sunday.

**Pull request merged**